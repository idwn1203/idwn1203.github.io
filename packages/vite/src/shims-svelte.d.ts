
declare module '@sveltejs/kit/vite' {
	import { Plugin } from 'vite';

	/**
	 * Returns the SvelteKit Vite plugins.
	 */
	export function sveltekit(): Promise<Plugin[]>;
	export { vitePreprocess } from '@sveltejs/vite-plugin-svelte';
}
