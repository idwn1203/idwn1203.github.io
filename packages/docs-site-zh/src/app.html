<!DOCTYPE html>
<html lang="zh-CN">
	<head>
		<meta charset="utf-8" />
		<meta name="viewport" content="width=device-width, initial-scale=1" />
		<meta property="og:type" content="website">
		<meta property="og:title" content="Sveltepress">
		<meta property="og:url" content="https://sveltepress.site">
		<meta property="og:description" content="一个简单易用的站点构建工具">
		<meta property="og:image" content="/android-chrome-512x512.png" />
		<meta property="og:image:width" content="512">
		<meta property="og:image:height" content="512">
		<link rel="icon" href="%sveltekit.assets%/sveltepress.svg" type="image/svg+xml" />
		<link rel="alternate icon" href="/favicon.ico" />
		<link rel="apple-touch-icon" href="/apple-touch-icon.png">
		<link rel="mask-icon" href="/sveltepress.svg" color="#ffffff">
		<meta name="msapplication-TileColor" content="#ffffff">
		<link rel="shortcut icon" type="image/png" href="/android-chrome-192x192.png">
		%sveltekit.head%
	</head>
	<body data-sveltekit-preload-data="hover">
		<div style="display: contents">%sveltekit.body%</div>
	</body>
</html>
