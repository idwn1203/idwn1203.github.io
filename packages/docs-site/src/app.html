<!DOCTYPE html>
<html lang="en">
	<head>
		<meta charset="utf-8" />
		<meta name="viewport" content="width=device-width, initial-scale=1" />
		<meta property="og:type" content="website">
		<meta property="og:title" content="Sveltepress">
		<meta property="og:url" content="https://sveltepress.site">
		<meta property="og:description" content="A simple, easy to use, content centered site build tool with the full power of Sveltekit">
		<meta property="og:image" content="https://sveltepress.site/android-chrome-512x512.png" />
		<meta property="og:image:width" content="512">
		<meta property="og:image:height" content="512">
		<link rel="icon" href="%sveltekit.assets%/sveltepress.svg" type="image/svg+xml" />
		<link rel="alternate icon" href="/favicon.ico" />
		<link rel="apple-touch-icon" href="/apple-touch-icon.png">
		<link rel="mask-icon" href="/sveltepress.svg" color="#ffffff">
		<meta name="msapplication-TileColor" content="#ffffff">
		<link rel="shortcut icon" type="image/png" href="/android-chrome-192x192.png">
		%sveltekit.head%
	</head>
	<body data-sveltekit-preload-data="hover">
		<div style="display: contents">%sveltekit.body%</div>
	</body>
</html>
