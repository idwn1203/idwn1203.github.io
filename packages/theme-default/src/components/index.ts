import Floating from '@sveltepress/twoslash/FloatingWrapper.svelte'
import Expansion from './Expansion.svelte'
import Link from './Link.svelte'
import Tabs from './Tabs.svelte'
import TabPanel from './TabPanel.svelte'
import InstallPkg from './InstallPkg.svelte'
import CopyCode from './CopyCode.svelte'
import IconifyIcon from './IconifyIcon.svelte'

export {
  Floating,
  IconifyIcon,
  CopyCode,
  Expansion,
  Link,
  Tabs,
  TabPanel,
  InstallPkg,
}
