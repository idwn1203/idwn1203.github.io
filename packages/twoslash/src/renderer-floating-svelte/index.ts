import { type RendererRichOptions, type TwoslashRenderer, rendererRich } from '@shikijs/twoslash'
import type { Element, ElementContent, Text } from 'hast'
import { gfmFromMarkdown } from 'mdast-util-gfm'
import { fromMarkdown } from 'mdast-util-from-markdown'
import { defaultHandlers, toHast } from 'mdast-util-to-hast'
import type { ShikiTransformerContextCommon } from 'shiki'

interface RendererFloatingSvelteOptions extends RendererRichOptions {}

function rendererFloatingSvelte(options: RendererFloatingSvelteOptions = {}): TwoslashRenderer {
  const rich = rendererRich({
    ...options,
    renderMarkdown,
    renderMarkdownInline,
    hast: {
      hoverToken: {
        tagName: 'Floating',
      },
      hoverCompose: compose,
      queryToken: {
        tagName: 'Floating',
        properties: {
          alwaysShow: '{true}',
        },
      },
      queryCompose: compose,
      completionCompose({ popup, cursor }) {
        return [
          <Element>{
            type: 'element',
            tagName: 'Floating',
            properties: {
              alwaysShow: '{true}',
            },
            children: [
              cursor,
              {
                type: 'element',
                tagName: 'svelte:fragment',
                properties: {
                  slot: 'floating-content',
                },
                children: [popup],
              },
            ],
          },
        ]
      },
    },
  })
  return rich
}

function compose(parts: { token: Element | Text; popup: Element }): ElementContent[] {
  return [
    {
      type: 'element',
      tagName: 'span',
      properties: {},
      children: [parts.token as any],
    },
    {
      type: 'element',
      tagName: 'svelte:fragment',
      properties: {
        slot: 'floating-content',
      },
      children: [parts.popup],
    },
  ]
}

function renderMarkdown(this: ShikiTransformerContextCommon, md: string): ElementContent[] {
  const mdast = fromMarkdown(
    md.replace(/{@link ([^}]*)}/g, '$1'), // replace jsdoc links
    { mdastExtensions: [gfmFromMarkdown()] },
  )

  return (toHast(
    mdast,
    {
      handlers: {
        code: (state, node) => {
          const lang = node.lang || ''
          if (lang) {
            return this.codeToHast(
              node.value,
              {
                ...this.options,
                transformers: [],
                lang,
              },
            ).children[0] as Element
          }
          return defaultHandlers.code(state, node)
        },
      },
    },
  ) as Element).children
}

function renderMarkdownInline(this: ShikiTransformerContextCommon, md: string, context?: string): ElementContent[] {
  if (context === 'tag:param')
    md = md.replace(/^([\w$-]+)/, '`$1` ')

  const children = renderMarkdown.call(this, md)
  if (children.length === 1 && children[0].type === 'element' && children[0].tagName === 'p')
    return children[0].children
  return children
}

export default rendererFloatingSvelte
