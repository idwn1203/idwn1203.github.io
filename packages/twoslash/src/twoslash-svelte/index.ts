import { readFileSync } from 'node:fs'
import { resolve } from 'node:path'
import { fileURLToPath } from 'node:url'
import type { CreateTwoslashOptions, TwoslashExecuteOptions, TwoslashInstance, TwoslashNode } from 'twoslash'
import { createTwoslasher as createTwoslasherBase } from 'twoslash'
import { DocumentSnapshot } from 'svelte-language-server/dist/src/plugins/typescript/DocumentSnapshot.js'
import { Document } from 'svelte-language-server/dist/src/lib/documents/index.js'
import { parse } from 'svelte/compiler'

const dirname = fileURLToPath(new URL('.', import.meta.url))

const additionalTypes = readFileSync(resolve(dirname, 'types-template.d.ts'), 'utf-8')

export interface CreateTwoslashSvelteOptions extends CreateTwoslashOptions {
  customTypes?: string
}

export interface TwoslashSvelteExecuteOptions extends TwoslashExecuteOptions {
}

export async function createTwoslasher(createTwoslashSvelteOptions: CreateTwoslashSvelteOptions = {}): Promise<TwoslashInstance> {
  const base = await createTwoslasherBase({
    ...createTwoslashSvelteOptions,
  })
  function twoslasher(code: string, extension?: string, options: TwoslashSvelteExecuteOptions = {}) {
    if (extension !== 'svelte')
      return base(code, extension, options)
    const svelteDoc = new Document('index.svelte', code)

    const tsxDoc = DocumentSnapshot.fromDocument(svelteDoc, {
      parse,
      version: '4',
      transformOnTemplateError: false,
      typingsNamespace: '',
    })

    const twoslashReturn = base([tsxDoc.getFullText(), additionalTypes].join('\n'), 'tsx', {
      compilerOptions: {
        jsx: 1,
        types: ['@sveltepress/vite/types', '@sveltepress/theme-default/types'],
      },
      shouldGetHoverInfo(identifier) {
        return !['__sveltets', 'Index__', 'svelteHTML', 'render', '$$_', 'target', 'props'].some(id => identifier.startsWith(id))
      },
    })

    twoslashReturn.meta.extension = 'svelte'

    function mapNode<T extends TwoslashNode>(node: T) {
      const { line, character } = tsxDoc.getOriginalPosition(node)
      node.line = line
      node.character = character
      node.start = svelteDoc.offsetAt(node)
      return node
    }

    twoslashReturn.code = code
    twoslashReturn.hovers.forEach(mapNode)
    twoslashReturn.nodes = twoslashReturn.nodes.filter(n => n.line > -1)
    return twoslashReturn
  }

  twoslasher.getCacheMap = base.getCacheMap

  return twoslasher
}
