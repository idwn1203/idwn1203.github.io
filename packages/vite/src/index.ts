import type { PluginOption } from 'vite'
import { sveltekit } from '@sveltejs/kit/vite'
import vitePluginInspect from 'vite-plugin-inspect'
import SveltepressVitePlugin from './plugin.js'
import type { Highlighter, LoadTheme, ResolvedTheme, SiteConfig, SveltepressVitePluginOptions } from './types'
import mdToSvelte from './markdown/md-to-svelte'

export * as log from './utils/log.js'

const sveltepress: (options: SveltepressVitePluginOptions) => PluginOption = ({
  theme,
  addInspect,
  siteConfig,
  remarkPlugins,
  rehypePlugins,
} = {
  addInspect: false,
}) => {
  const requiredSiteConfig: Required<SiteConfig> = {
    title: siteConfig?.title || 'Untitled site',
    description: siteConfig?.description || 'Build by Sveltepress',
  }
  const corePlugin = [
    SveltepressVitePlugin({
      theme,
      siteConfig: requiredSiteConfig,
      remarkPlugins,
      rehypePlugins,
    }),
    sveltekit(),
  ]

  const plugins = typeof theme?.vitePlugins === 'function'
    ? theme.vitePlugins(corePlugin)
    : [
        theme?.vitePlugins,
        ...corePlugin,
      ]
  if (addInspect)
    plugins.unshift(vitePluginInspect())

  return plugins
}

export { sveltepress, mdToSvelte }
export type { LoadTheme, Highlighter, SiteConfig, SveltepressVitePluginOptions, ResolvedTheme }
