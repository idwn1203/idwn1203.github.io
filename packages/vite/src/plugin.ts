import { resolve } from 'node:path'
import { existsSync, mkdirSync } from 'node:fs'
import process from 'node:process'
import type { PluginOption } from 'vite'

import type { Plugin } from 'unified'
import type { SveltepressVitePluginOptions } from './types'
import { wrapPage } from './utils/wrap-page.js'

export const BASE_PATH = resolve(process.cwd(), '.sveltepress')

// virtual modules
const SVELTEPRESS_SITE_CONFIG_MODULE = 'virtual:sveltepress/site'

// only the src/routes/**/*.+(page|layout).(svelte|md) will need to be wrapped by theme.pageLayout
export const PAGE_OR_LAYOUT_RE = /\/src\/routes(\/[\(\)\[\]\w- ]+)*\/\+(?:page|layout)(@[\w-]*)?\.(?:svelte|md)$/

if (!existsSync(BASE_PATH))
  mkdirSync(BASE_PATH, { recursive: true })

const sveltepress: (options: SveltepressVitePluginOptions) => PluginOption = ({
  theme,
  siteConfig,
  rehypePlugins,
  remarkPlugins,
}) => {
  const allRemarkPlugins: Plugin[] = []
  const allRehypePlugins: Plugin[] = []

  if (theme?.remarkPlugins)
    allRemarkPlugins.push(...theme.remarkPlugins)
  if (remarkPlugins)
    allRemarkPlugins.push(...remarkPlugins)
  if (theme?.rehypePlugins)
    allRehypePlugins.push(...theme.rehypePlugins)
  if (rehypePlugins)
    allRehypePlugins.push(...rehypePlugins)

  function getLayout(path: string) {
    let layout: string | undefined
    if (isRootLayout(path))
      layout = theme?.globalLayout
    else if (isPage(path))
      layout = theme?.pageLayout
    return layout
  }

  const getWrappedCode = async (id: string, src: string) => (await wrapPage({
    id,
    mdOrSvelteCode: src,
    ...theme,
    remarkPlugins: allRemarkPlugins,
    rehypePlugins: allRehypePlugins,
    layout: getLayout(id),
  })).wrappedCode

  return {
    name: '@sveltepress/vite',
    /**
     * Must enable this because vite-plugin-svelte enabled this too
     * @see https://github.com/sveltejs/vite-plugin-svelte/blob/1cef575c8f9188456934e38dad7a869b43fe7d46/packages/vite-plugin-svelte/src/index.ts#L58
     */
    enforce: 'pre',
    config: () => ({
      server: {
        fs: {
          allow: ['.sveltepress'],
        },
      },
      resolve: {
        alias: {
          $sveltepress: resolve(process.cwd(), '.sveltepress'),
        },
      },
    }),
    resolveId(id) {
      if (id === SVELTEPRESS_SITE_CONFIG_MODULE)
        return SVELTEPRESS_SITE_CONFIG_MODULE
    },
    load(id) {
      if (id === SVELTEPRESS_SITE_CONFIG_MODULE)
        return `export default ${JSON.stringify(siteConfig)}`
    },
    async transform(src, id) {
      if (PAGE_OR_LAYOUT_RE.test(id)) {
        const code = await getWrappedCode(id, src)
        return code
      }
    },
    async handleHotUpdate(ctx) {
      const { file } = ctx
      if (PAGE_OR_LAYOUT_RE.test(file)) {
        const src = await ctx.read()
        // overwrite read() to return content parsed by md-to-svelte so that sveltekit can handle the HMR
        ctx.read = async () => await getWrappedCode(file, src)
      }
    },
  }
}

function isPage(path: string) {
  return path.endsWith('+page.svelte') || path.endsWith('+page.md')
}

function isRootLayout(path: string) {
  return path.endsWith('src/routes/+layout.svelte') || path.endsWith('src/routes/+layout.md')
}

export default sveltepress
