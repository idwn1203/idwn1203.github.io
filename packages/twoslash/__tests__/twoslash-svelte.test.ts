import { readFileSync } from 'node:fs'
import { resolve } from 'node:path'
import { describe, expect, it } from 'vitest'
import { createTwoslasher } from '../src'

const svelteCode = readFileSync(resolve(import.meta.dirname, 'test.svelte'), 'utf-8')

const twoslasher = await createTwoslasher()

describe('shiki', () => {
  it('twoslash svelte hover', async () => {
    expect(twoslasher(svelteCode, 'svelte').hovers).toMatchSnapshot()
  })
})
