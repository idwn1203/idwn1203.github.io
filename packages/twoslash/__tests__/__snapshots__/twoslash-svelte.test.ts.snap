// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`shiki > twoslash svelte hover 1`] = `
[
  {
    "character": 15,
    "docs": undefined,
    "length": 3,
    "line": 1,
    "start": 51,
    "tags": undefined,
    "target": "FOO",
    "text": "const FOO: "BAR"",
    "type": "hover",
  },
  {
    "character": 11,
    "docs": "The \`onMount\` function schedules a callback to run as soon as the component has been mounted to the DOM.
It must be called during the component's initialisation (but doesn't need to live *inside* the component;
it can be called from an external module).

If a function is returned _synchronously_ from \`onMount\`, it will be called when the component is unmounted.

\`onMount\` does not run inside a [server-side component](https://svelte.dev/docs#run-time-server-side-component-api).

https://svelte.dev/docs/svelte#onmount",
    "length": 7,
    "line": 5,
    "start": 104,
    "tags": undefined,
    "target": "onMount",
    "text": "(alias) function onMount<T>(fn: () => NotFunction<T> | Promise<NotFunction<T>> | (() => any)): void
import onMount",
    "type": "hover",
  },
  {
    "character": 9,
    "docs": "Base class for Svelte components with some minor dev-enhancements. Used when dev=true.

Can be used to create strongly typed Svelte components.

#### Example:

You have component library on npm called \`component-library\`, from which
you export a component called \`MyComponent\`. For Svelte+TypeScript users,
you want to provide typings. Therefore you create a \`index.d.ts\`:
\`\`\`ts
import { SvelteComponent } from "svelte";
export class MyComponent extends SvelteComponent<{foo: string}> {}
\`\`\`
Typing this makes it possible for IDEs like VS Code with the Svelte extension
to provide intellisense and to use the component like this in a Svelte file
with TypeScript:
\`\`\`svelte
<script lang="ts">
	import { MyComponent } from "component-library";
</script>
<MyComponent foo={'bar'} />
\`\`\`",
    "length": 15,
    "line": 6,
    "start": 137,
    "tags": undefined,
    "target": "FloatingWrapper",
    "text": "(alias) class FloatingWrapper<Props extends Record<string, any> = any, Events extends Record<string, any> = any, Slots extends Record<string, any> = any>
import FloatingWrapper",
    "type": "hover",
  },
  {
    "character": 6,
    "docs": undefined,
    "length": 5,
    "line": 8,
    "start": 208,
    "tags": undefined,
    "target": "count",
    "text": "let count: number",
    "type": "hover",
  },
  {
    "character": 2,
    "docs": "The \`onMount\` function schedules a callback to run as soon as the component has been mounted to the DOM.
It must be called during the component's initialisation (but doesn't need to live *inside* the component;
it can be called from an external module).

If a function is returned _synchronously_ from \`onMount\`, it will be called when the component is unmounted.

\`onMount\` does not run inside a [server-side component](https://svelte.dev/docs#run-time-server-side-component-api).

https://svelte.dev/docs/svelte#onmount",
    "length": 7,
    "line": 10,
    "start": 221,
    "tags": undefined,
    "target": "onMount",
    "text": "(alias) onMount<void>(fn: () => void | (() => any) | Promise<void>): void
import onMount",
    "type": "hover",
  },
  {
    "character": 4,
    "docs": undefined,
    "length": 7,
    "line": 11,
    "start": 241,
    "tags": undefined,
    "target": "console",
    "text": "var console: Console",
    "type": "hover",
  },
  {
    "character": 12,
    "docs": "[MDN Reference](https://developer.mozilla.org/docs/Web/API/console/log)",
    "length": 3,
    "line": 11,
    "start": 249,
    "tags": undefined,
    "target": "log",
    "text": "(method) Console.log(...data: any[]): void",
    "type": "hover",
  },
  {
    "character": 0,
    "docs": undefined,
    "length": 13,
    "line": 15,
    "start": 280,
    "tags": undefined,
    "target": "createElement",
    "text": "any",
    "type": "hover",
  },
  {
    "character": 24,
    "docs": undefined,
    "length": 5,
    "line": 15,
    "start": 304,
    "tags": undefined,
    "target": "count",
    "text": "let count: number",
    "type": "hover",
  },
  {
    "character": 13,
    "docs": undefined,
    "length": 5,
    "line": 16,
    "start": 327,
    "tags": undefined,
    "target": "count",
    "text": "let count: number",
    "type": "hover",
  },
  {
    "character": 1,
    "docs": "Base class for Svelte components with some minor dev-enhancements. Used when dev=true.

Can be used to create strongly typed Svelte components.

#### Example:

You have component library on npm called \`component-library\`, from which
you export a component called \`MyComponent\`. For Svelte+TypeScript users,
you want to provide typings. Therefore you create a \`index.d.ts\`:
\`\`\`ts
import { SvelteComponent } from "svelte";
export class MyComponent extends SvelteComponent<{foo: string}> {}
\`\`\`
Typing this makes it possible for IDEs like VS Code with the Svelte extension
to provide intellisense and to use the component like this in a Svelte file
with TypeScript:
\`\`\`svelte
<script lang="ts">
	import { MyComponent } from "component-library";
</script>
<MyComponent foo={'bar'} />
\`\`\`",
    "length": 15,
    "line": 19,
    "start": 346,
    "tags": undefined,
    "target": "FloatingWrapper",
    "text": "(alias) class FloatingWrapper<Props extends Record<string, any> = any, Events extends Record<string, any> = any, Slots extends Record<string, any> = any>
import FloatingWrapper",
    "type": "hover",
  },
  {
    "character": 26,
    "docs": "Base class for Svelte components with some minor dev-enhancements. Used when dev=true.

Can be used to create strongly typed Svelte components.

#### Example:

You have component library on npm called \`component-library\`, from which
you export a component called \`MyComponent\`. For Svelte+TypeScript users,
you want to provide typings. Therefore you create a \`index.d.ts\`:
\`\`\`ts
import { SvelteComponent } from "svelte";
export class MyComponent extends SvelteComponent<{foo: string}> {}
\`\`\`
Typing this makes it possible for IDEs like VS Code with the Svelte extension
to provide intellisense and to use the component like this in a Svelte file
with TypeScript:
\`\`\`svelte
<script lang="ts">
	import { MyComponent } from "component-library";
</script>
<MyComponent foo={'bar'} />
\`\`\`",
    "length": 15,
    "line": 20,
    "start": 426,
    "tags": undefined,
    "target": "FloatingWrapper",
    "text": "(alias) class FloatingWrapper<Props extends Record<string, any> = any, Events extends Record<string, any> = any, Slots extends Record<string, any> = any>
import FloatingWrapper",
    "type": "hover",
  },
]
`;
